richtextfx/src/main/java/org/fxmisc/richtext/GenericStyledArea.java

```diff
@@ -406,42 +407,72 @@ protected void invalidated() {
      *                                                                        *
      * ********************************************************************** */
 
-    private final ObjectProperty<Consumer<MouseEvent>> onOutsideSelectionMousePress = new SimpleObjectProperty<>(e -> {
-        CharacterHit hit = hit(e.getX(), e.getY());
-        moveTo(hit.getInsertionIndex(), SelectionPolicy.CLEAR);
-    });
+    @Deprecated private final ObjectProperty<Consumer<MouseEvent>> onOutsideSelectionMousePress = new SimpleObjectProperty<>( null );
     @Override public final ObjectProperty<Consumer<MouseEvent>> onOutsideSelectionMousePressProperty() { return onOutsideSelectionMousePress; }
 
-    private final ObjectProperty<Consumer<MouseEvent>> onInsideSelectionMousePressRelease = new SimpleObjectProperty<>(e -> {
-        CharacterHit hit = hit(e.getX(), e.getY());
-        moveTo(hit.getInsertionIndex(), SelectionPolicy.CLEAR);
+    @Override public final EventHandler<MouseEvent> getOnOutsideSelectionMousePressed() { return onOutsideSelectionMousePressed.get(); }
+    @Override public final void setOnOutsideSelectionMousePressed(EventHandler<MouseEvent> handler) { onOutsideSelectionMousePressed.set( handler ); }
+    @Override public final ObjectProperty<EventHandler<MouseEvent>> onOutsideSelectionMousePressedProperty() { return onOutsideSelectionMousePressed; }
+    private final ObjectProperty<EventHandler<MouseEvent>> onOutsideSelectionMousePressed = new SimpleObjectProperty<>( e -> 
+    {
+        if ( onOutsideSelectionMousePress.get() != null )  onOutsideSelectionMousePress.get().accept(e);
+        else {
+            CharacterHit hit = hit(e.getX(), e.getY());
+            moveTo(hit.getInsertionIndex(), SelectionPolicy.CLEAR);
+        }
     });
```

Comment:
I decided to move this conversation down a bit (delete originals and re-add them here) so that the full scope of the code could be more easily viewed in GitHub.

Originally, I wrote:
> Could you move all deprecated methods down to the bottom of the file? It gets them out of the way.

You responded with: 
> Tried to do this, but then the compiler complains about an Illegal forward reference so I've reverted this back to the way it was. [Confused emoticon]

So, let's break down what's going on here...
````java
// old method & name
@Deprecated private final ObjectProperty<Consumer<MouseEvent>> onOutsideSelectionMousePress 
 = new SimpleObjectProperty<>( null );
@Override public final ObjectProperty<Consumer<MouseEvent>> onOutsideSelectionMousePressProperty() { 
 return onOutsideSelectionMousePress; 
}
// new method & name
private final ObjectProperty<EventHandler<MouseEvent>> onOutsideSelectionMousePressed 
 = new SimpleObjectProperty<>( e -> 
{
    // in order to support the deprecated method `onOutsideSelectionMousePressProperty` and
    // their respective getter/setter methods, you need to run that method first if it's not null 
    if ( onOutsideSelectionMousePress.get() != null )  onOutsideSelectionMousePress.get().accept(e);
    else {
    // If it's null, then run the default behavior specified here
        CharacterHit hit = hit(e.getX(), e.getY());
        moveTo(hit.getInsertionIndex(), SelectionPolicy.CLEAR);
    }
});
````
Since you reference the `onOutsideSelectionMousePress` field directly rather than through its method, the field needs to be declared before its first reference. This explains your "illegal forward reference" issue since you put it at the bottom of the file. If you put the field declaration and its corresponding property method at the bottom of the file, (with the field's declaration occurring one line above its method), then the compiler won't raise that issue if you refer to the field not directly but indirectly through its property method.

However, I have to ask whether your approach is the best in terms of modifying the source code as little as possible. If a developer overrides the original `press` property, then the `press` property still gets run because the code actually runs the `pressed` property, which will then run the `press` property because it's non-null. If a developer overrides the new `pressed` property, then it gets rid of the null check and runs whatever custom code the developer specifies. So why not write it this way:
````java
// Place this old code at the bottom of the file
// I only put it here to communicate my point easier
private final ObjectProperty<Consumer<MouseEvent>> onOutsideSelectionMousePress 
= new SimpleObjectProperty<>(
    // Add this comment to the source. No need to change anything here.
    // Note: this code should be moved `onOutsideSelectionMousePressed
    // in the next major release before removing this deprecated field
    // due to a change in type (Consumer to EventHandler)
    e -> {
        CharacterHit hit = hit(e.getX(), e.getY());
        moveTo(hit.getInsertionIndex(), SelectionPolicy.CLEAR);
});
@Deprecated 
@Override public final ObjectProperty<Consumer<MouseEvent>> onOutsideSelectionMousePressProperty() { 
return onOutsideSelectionMousePress; 
}
// new method & name
private final ObjectProperty<EventHandler<MouseEvent>> onOutsideSelectionMousePressed 
 = new SimpleObjectProperty<>( e -> 
{
    // in order to support the deprecated method `onOutsideSelectionMousePressProperty` and
    // their respective getter/setter methods, this handler just runs the property's code
    // which is referenced through its property method, and not directly to avoid the illegal forward reference
    onOutsideSelectionMousePressProperty().get().accept(e);
});
````
Also, you deprecated the field, which was private and inaccessible anyway, but not the field's public property method, which is accessible and needs to be deprecated. I fixed that in my example above.

Reply:
I understand the need to "modify the source code as little as possible" and have no problem with your alternative.

I deprecated the private field for internal "documentation" purposes, and the public property wasn't because it's already deprecated in the ViewActions interface. :-)